tough/src/transport.rs

```diff
@@ -1,36 +1,120 @@
-pub trait Transport {
-    /// The type of `Read` object that the `fetch` function will return.
-    type Stream: Read;
+///
+/// The trait hides the underlying types involved by returning the `Read` object as a
+/// `Box<dyn Read + Send>` and by requiring concrete type [`TransportError`] as the error type.
+///
+pub trait Transport: Debug + DynClone {
+    /// Opens a `Read` object for the file specified by `url`.
+    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError>;
+}
 
-    /// The type of error that the `fetch` function will return.
-    type Error: std::error::Error + Send + Sync + 'static;
+// Implement `Clone` for `Transport` trait objects.
+dyn_clone::clone_trait_object!(Transport);
 
-    /// Opens a `Read` object for the file specified by `url`.
-    fn fetch(&self, url: Url) -> Result<Self::Stream, Self::Error>;
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// The kind of error that the transport object experienced during `fetch`.
+///
+/// # Why
+///
+/// Some TUF operations need to know if a [`Transport`] failure is a result of a file not being
+/// found. In particular:
+/// > 5.1.2. Try downloading version N+1 of the root metadata file `[...]` If this file is not
+/// > available `[...]` then go to step 5.1.9.
+///
+/// To distinguish this case from other [`Transport`] failures, we use
+/// `TransportErrorKind::FileNotFound`.
+///
+#[derive(Debug, Copy, Clone)]
+#[non_exhaustive]
+pub enum TransportErrorKind {
+    /// The trait does not handle the URL scheme named in `String`. e.g. `file://` or `http://`.
+    UnsupportedUrlScheme,
+    /// The file cannot be found.
+    FileNotFound,
+    /// The transport failed for any other reason, e.g. IO error, HTTP broken pipe, etc.
+    Failure,
+}
+
+/// The error type that [`Transport`] `fetch` returns.
+#[derive(Debug)]
+pub struct TransportError {
+    /// The kind of error that occurred.
+    pub kind: TransportErrorKind,
+    /// The URL that the transport was trying to fetch.
+    pub url: String,
+    /// The underlying error that occurred.
+    pub source: Box<dyn std::error::Error + Send + Sync>,
+}
+
+impl TransportError {
+    /// Creates a new [`TransportError`].
+    pub fn new<S, E>(kind: TransportErrorKind, url: S, source_error: E) -> Self
+    where
+        E: Into<Box<dyn std::error::Error + Send + Sync>>,
+        S: AsRef<str>,
+    {
+        Self {
+            kind,
+            url: url.as_ref().to_owned(),
+            source: source_error.into(),
+        }
+    }
+
+    /// Creates a [`TransportError`] for reporting an unhandled URL type.
+    pub fn unsupported_url<S: AsRef<str>>(url: S) -> Self {
+        TransportError::new(
+            TransportErrorKind::UnsupportedUrlScheme,
+            url,
+            "Transport cannot handle the given URL scheme.".to_string(),
+        )
+    }
+}
+
+/// [`TransportError`] implements the standard error interface.
```

Comment:
snafu 0.6.9 supports deriving Snafu on structs (not just for making an error opaque) and it might get you what you want here, without manually deriving Error and Display.

Reply:
Addressed in [59899dd](https://github.com/awslabs/tough/compare/3e858e5b5d629576a0c13f3f7b627cd8c49af9a4..59899ddf735ceb7e4edc883a9b1c6a195506f820),